File: classes/ImageDecoder.php
<?php
class ImageDecoder
{
    private $debug = false;
    private $width = false;
    private $height = false;
    private $img_code = false;
    private $layers = [];
    private $pic = [];
    private $pic_pos = 0;


    public function __construct($width = false, $height = false, $img_code = false, $debug = false)
    {
        if ($width !== false) {
            $this->setWidth($width);
        }
        if ($height !== false) {
            $this->setHeight($height);
        }
        if ($img_code !== false) {
            $this->setImgCode($img_code);
        }

        $this->setDebug($debug);

        if ($this->checkIsValidInputs()) {
            $this->setImageLayers();
        }

        return true;
    }

    public function setWidth($width)
    {
        $this->width = $width;

        if ($this->checkIsValidInputs()) {
            $this->setImageLayers();
        }
        return true;
    }

    public function setHeight($height)
    {
        $this->height = $height;

        if ($this->checkIsValidInputs()) {
            $this->setImageLayers();
        }
        return true;
    }

    public function setImgCode($img_code)
    {
        $this->img_code = $img_code;

        if ($this->checkIsValidInputs()) {
            $this->setImageLayers();
        }
        return true;
    }


    public function setDebug($debug = false)
    {
        $this->debug = $debug;
        return true;
    }

    public function checkIsValidInputs()
    {
        if ($this->width !== false && $this->height !== false && $this->img_code !== false) {
            return true;
        } else {
            return false;
        }
    }

    public function setImageLayers()
    {
        $layer = 0;
        $this->layers = [];
        $this->pic = [];
        $this->pic_pos = 0;

        $img_code = preg_split('//', $this->img_code, -1, PREG_SPLIT_NO_EMPTY);
        $debug_info = '';

        for ($pos = 0; $pos < count($img_code); $pos++) {
            if ($pos > 0 && $pos % $this->width == 0) {
                $debug_info .= "\n";
            }
            if ($pos % ($this->width * $this->height) == 0) {
                if ($pos > 0) {
                    $layer++;
                }
                $debug_info .= "\nLayer $layer:\n";
                $this->layers[$layer] = [0, 0, 0];
            }

            $this->layers[$layer][intval($img_code[$pos])]++;
            $debug_info .= $img_code[$pos];

            $this->pic_pos = $pos % ($this->width * $this->height);
            if ($img_code[$pos] != 2 && !isset($this->pic[$this->pic_pos])) {
                $this->pic[$this->pic_pos] = $img_code[$pos];
            }
        }

        $debug_info .= "\n";

        if ($this->debug) {
            echo $debug_info;
        }
    }

    private static function sortByFewest0($a, $b)
    {
        if ($a[0] == $b[0]) {
            return 0;
        } else {
            return $a[0] < $b[0] ? -1 : 1;
        }
    }

    public function getFewest0Layer()
    {
        usort($this->layers, array('ImageDecoder', 'sortByFewest0'));
        return $this->layers[0];
    }

    private function setPixel($pixel, $raw = false)
    {
        return ($raw) ? $pixel : ($pixel == 1 ? "\033[47m \033[0m" : ' ');
    }

    public function drawPic($return = false, $raw = false)
    {
        $output = '';
        $output .= "\n";
        for ($pos = 0; $pos < count($this->pic); $pos++) {
            if ($pos > 0 && $pos % $this->width == 0) {
                $output .= "\n";
            }
            $output .= $this->setPixel($this->pic[$pos], $raw);
        }
        $output .= "\n";

        if ($return) {
            return trim($output);
        } else {
            echo $output;
        }
    }
}

File: machines/image_decoder.py
from typing import Dict


class ImageDecoder:
    __debug: bool = False
    __width: int = None
    __height: int = None
    __img_code: str = None
    __layers: Dict = None
    __pic: Dict = None
    __pic_pos: int = None

    def __init__(self, width: int = None, height: int = None, img_code: str = None, debug: bool = False):
        if width is not None:
            self.set_width(width)
        if height is not None:
            self.set_height(height)
        if img_code is not None:
            self.set_img_code(img_code)

        self.set_debug(debug)

        if self.check_is_valid_inputs():
            self.set_image_layers()

    def set_width(self, width: int):
        self.__width = width
        if self.check_is_valid_inputs():
            self.set_image_layers()
        return True

    def set_height(self, height: int):
        self.__height = height
        if self.check_is_valid_inputs():
            self.set_image_layers()
        return True

    def set_img_code(self, img_code: str):
        self.__img_code = img_code
        if self.check_is_valid_inputs():
            self.set_image_layers()
        return True

    def set_debug(self, debug: bool = False):
        self.__debug = debug
        return True

    def check_is_valid_inputs(self):
        return True if (self.__width is not None and self.__height is not None and self.__img_code is not None) else False

    def set_image_layers(self):
        layer = 0
        self.__layers = {}
        self.__pic = {}
        self.__pic_pos = 0

        img_code = list(self.__img_code.replace(" ", ""))
        debug_info = ""

        for pos in range(len(img_code)):
            if pos > 0 and pos % self.__width == 0:
                debug_info += "\n"
            if pos % (self.__width * self.__height) == 0:
                if pos > 0:
                    layer += 1
                debug_info = f"\nLayer {layer}:\n"
                self.__layers[layer] = [0, 0, 0]
            self.__layers[layer][int(img_code[pos])] += 1
            debug_info += img_code[pos]

            self.__pic_pos = pos % (self.__width * self.__height)
            if img_code[pos] != "2" and self.__pic_pos not in self.__pic:
                self.__pic[self.__pic_pos] = img_code[pos]

        debug_info += "\n"

        if self.__debug:
            print(debug_info)

    @staticmethod
    def __sort_by_fewest_0(a, b):
        return 0 if a[0] == b[0] else -1 if a[0] < b[0] else 1

    def get_fewest_0_layer(self):
        self.__layers = sorted(self.__layers.items(), key=lambda item: item[1][0])
        return self.__layers[0][1]

    def __set_pixel(self, pixel, raw=False):
        return pixel if raw else "\033[47m \033[0m" if pixel == "1" else " "

    def drawPic(self, ret=False, raw=False):
        output = ""
        output += "\n"

        for pos in range(len(self.__pic)):
            if pos > 0 and pos % self.__width == 0:
                output += "\n"
            output += self.__set_pixel(self.__pic[pos], raw)
        output += "\n"

        if ret:
            return output.strip()
        else:
            print(output)
